### ISSUE TITLE
Recursive fs.readdir blocks browser window 

### ISSUE THREAD

**Issue by knoopx** [atom/atom-shell#1155]
Calling recursively `require('fs').readdir` within the renderer context blocks the browser window so it does not responding to mouse events or repaint.

I'm creating an mp3 player and I stumbled with this problem. I want to recursively scan for mp3 files under a directory and persist them. However whenever I click a button and the index function starts it blocks the UI until it finishes.

Can you explain why does this happen and how to workaround it? Thank you

**Comment by paulcbetts** [comment]
@knoopx Can you use a Web Worker to do this?

**Comment by knoopx** [comment]
according to https://github.com/atom/atom-shell/issues/797 , atom shell does not have node integration within web workers so I can't use them :(

I also tried process.nextTick, async module and setTimeout tricks without any success.

**Comment by knoopx** [comment]
and `require('remote').require('glob')(path)` neither does the trick, it blocks the app in a process-wide level instead of window one.

**Comment by YuriSolovyov** [comment]
you can try `child_process.fork`

**Comment by zcbenz** [comment]
Can you share how you recursively call `fs.readdir`?

**Comment by zcbenz** [comment]
I'm closing this unless there is more information.

**Comment by zz85** [comment]
I believe I've encounter similar behaviour with [an app I'm writing](https://github.com/zz85/space-radar-electron) to scan and visualize disk usage. The way I call [fs.readdir is a simple recursion](https://github.com/zz85/space-radar-electron/blob/master/js/du.js#L18) but if ran within the same renderer process would cause the UI to lock.

I'm not sure if this is causing by any blocking calls, but I suspect that the recursive readdir calls are too too fast that it starves events within the same process. My current workaround to run fs operations in a webview. File information is passed via localStorage or by serializing and deserializing thru file system (it's workable thru ipc, I did this for a speed hack).

Just sharing some of my experiences, and I wrote a little more about the [speed hacks in this issue](https://github.com/zz85/space-radar-electron/issues/6) getting UI not to lock as much as possible.
